File: stickerlibrary/src/main/java/com/lcw/library/stickerview/StickerManager.java
package com.lcw.library.stickerview;

import android.graphics.Bitmap;
import android.graphics.Matrix;

import java.util.ArrayList;
import java.util.List;

/**
 * M Abdul Basit
 * Create by: Abdul Basit
 * Date: 2025/8/19
 * Time: 9:44 AM
 */
public class StickerManager {

    private static volatile StickerManager mInstance;

    private List<Sticker> mStickerList = new ArrayList<>();

    public static StickerManager getInstance() {
        if (mInstance == null) {
            synchronized (StickerManager.class) {
                if (mInstance == null) {
                    mInstance = new StickerManager();
                }
            }
        }
        return mInstance;
    }

    public void addSticker(Sticker sticker) {
        mStickerList.add(sticker);
    }

    public List<Sticker> getStickerList() {
        return mStickerList;
    }

    /**
     * removeSticker
     *
     * @param sticker
     */
    public void removeSticker(Sticker sticker) {
        Bitmap bitmap = sticker.getBitmap();
        if (bitmap != null && bitmap.isRecycled()) {
            bitmap.recycle();
        }
        mStickerList.remove(sticker);

    }

    /**
     * removeAllSticker
     */
    public void removeAllSticker() {
        for (int i = 0; i < mStickerList.size(); i++) {
            Bitmap bitmap = mStickerList.get(i).getBitmap();
            if (bitmap != null && bitmap.isRecycled()) {
                bitmap.recycle();
            }
        }
        mStickerList.clear();
    }

    /**
     * setFocusSticker
     *
     * @param focusSticker
     */
    public void setFocusSticker(Sticker focusSticker) {
        for (int i = 0; i < mStickerList.size(); i++) {
            Sticker sticker = mStickerList.get(i);
            if (sticker == focusSticker) {
                sticker.setFocus(true);
            } else {
                sticker.setFocus(false);
            }
        }
    }

    /**
     * clearAllFocus
     */
    public void clearAllFocus() {
        for (int i = 0; i < mStickerList.size(); i++) {
            Sticker sticker = mStickerList.get(i);
            sticker.setFocus(false);
        }
    }

    /**
     * getSticker
     *
     * @param x
     * @param y
     * @return
     */
    public Sticker getSticker(float x, float y) {

        float[] dstPoints = new float[2];
        float[] srcPoints = new float[]{x, y};

        for (int i = mStickerList.size() - 1; i >= 0; i--) {
            Sticker sticker = mStickerList.get(i);
            Matrix matrix = new Matrix();
            sticker.getMatrix().invert(matrix);
            matrix.mapPoints(dstPoints, srcPoints);
            if (sticker.getStickerBitmapBound().contains(dstPoints[0], dstPoints[1])) {
                return sticker;
            }
        }
        return null;
    }

    /**
     * To get Delete Button View
     *
     * @param x
     * @param y
     * @return
     */
    public Sticker getDelButton(float x, float y) {

        float[] dstPoints = new float[2];
        float[] srcPoints = new float[]{x, y};

        for (int i = mStickerList.size() - 1; i >= 0; i--) {
            Sticker sticker = mStickerList.get(i);
            Matrix matrix = new Matrix();
            sticker.getMatrix().invert(matrix);
            matrix.mapPoints(dstPoints, srcPoints);
            if (sticker.getDelBitmapBound().contains(dstPoints[0], dstPoints[1])) {
                return sticker;
            }
        }
        return null;

    }


}

File: stickerlibrary/src/main/java/com/basit/library/stickerview/StickerManager.kt
package com.basit.library.stickerview

import android.graphics.Matrix
import com.basit.library.stickerview.StickerFactory.currentSticker
import com.basit.library.stickerview.StickerFactory.isStickerFocused
import kotlin.concurrent.Volatile

/**
 * M Abdul Basit
 * Create by: Abdul Basit
 * Date: 2025/8/19
 * Time: 9:44 AM
 */
class StickerManager {

    val stickerList: MutableList<Sticker> = ArrayList<Sticker>()

    fun addSticker(sticker: Sticker?) {
        stickerList.add(sticker!!)
    }

    /**
     * removeSticker
     *
     * @param sticker
     */
    fun removeSticker(sticker: Sticker) {
        val bitmap = sticker.bitmap
        if (bitmap.isRecycled) {
            bitmap.recycle()
        }
        stickerList.remove(sticker)
    }

    /**
     * removeAllSticker
     */
    fun removeAllSticker() {
        for (i in stickerList.indices) {
            val bitmap = stickerList[i].bitmap
            if (bitmap.isRecycled) {
                bitmap.recycle()
            }
        }
        stickerList.clear()
    }

    /**
     * setFocusSticker
     *
     * @param focusSticker
     */
    fun setFocusSticker(focusSticker: Sticker?) {
        for (i in stickerList.indices) {
            val sticker = stickerList[i]
            if (sticker == focusSticker) {
                sticker.isFocus = true
                isStickerFocused.postValue(true)
            } else {
                sticker.isFocus = false
//                isStickerFocused.postValue(false)
            }
        }

    }

    /**
     * clearAllFocus
     */
    fun clearAllFocus() {
        for (i in stickerList.indices) {
            val sticker = stickerList[i]
            sticker.isFocus = false
            isStickerFocused.postValue(false)
        }
    }

    /**
     * getSticker
     *
     * @param x
     * @param y
     * @return
     */
    fun getSticker(x: Float, y: Float): Sticker? {
        val dstPoints = FloatArray(2)
        val srcPoints = floatArrayOf(x, y)

        for (i in stickerList.indices.reversed()) {
            val sticker = stickerList[i]
            val matrix = Matrix()
            sticker.matrix.invert(matrix)
            matrix.mapPoints(dstPoints, srcPoints)
            if (sticker.stickerBitmapBound!!.contains(dstPoints[0], dstPoints[1])) {
                return sticker
            }
        }
        return null
    }

    /**
     * To get Delete Button View
     *
     * @param x
     * @param y
     * @return
     */
    fun getDelButton(x: Float, y: Float): Sticker? {
        val dstPoints = FloatArray(2)
        val srcPoints = floatArrayOf(x, y)

        for (i in stickerList.indices.reversed()) {
            val sticker = stickerList[i]
            val matrix = Matrix()
            sticker.matrix.invert(matrix)
            matrix.mapPoints(dstPoints, srcPoints)
            if (sticker.delBitmapBound!!.contains(dstPoints[0], dstPoints[1])) {
                return sticker
            }
        }
        return null
    }
    /**
     * Get the currently focused sticker
     */
    fun getFocusSticker(): Sticker? {
        return stickerList.firstOrNull { it.isFocus }
    }


    companion object {
        @Volatile
        private var instance: StickerManager? = null

        fun getInstance(): StickerManager {
            return instance ?: synchronized(this) {
                instance ?: StickerManager().also { instance = it }
            }
        }
    }
}
